### PR TITLE
WoW64 Fixes & TEB Retrieval 

### PR DESCRIPTION
**Pull request by Zhentar**
A bunch of fixes & enhancements for WoW64, primarily focused around making sure the right ntdll module is used for resolving types.

$teb and $peb use the DbgEng provided module for symbol lookup, which does seem to be accurate from my testing.

For most everything else, we need to determine which ntdll module has which bitness. DbgEng knows which is which but of course does not simply expose it in a straightforward way. I ended up settling on `GetSymbolModuleWide( "${$ntnsym}!"` (and `$ntwsym`) to coax it out; it gives us the right answer in a single DbgEng call without having to parse anything and doesn't rely on any undocumented behavior (I think). Conveniently, it also matches the semantics that I think make the most sense (`$ntnsym` always returns a value; 64-bit when there is one and 32-bit otherwise, while `$ntwsym` always returns the 32-bit module when there is one and fails otherwise).

Since I needed a way to test those lookups, I added handling for DbgEng's nt! module shortcut, and invented a matching nt32! module shortcut.

I also threw in a `MemoryMarshal.Read<>` inspired `ReadMemAs<>` for easy & efficient reading of a single value (except implemented in the C++/CLI layer so that I never have to *say* I'm being `unsafe`).

### CODE REVIEW

**Comment by Zhentar** [comment]
No rush - I've hit the magical customer dump of edge cases that breaks everything it touches, probably better to give it a couple more days to see if any more of those things are in DbgShell 😆

**Comment by jazzdelightsme** [comment]
I'm curious: why are you interested in detecting wow64 targets?

File: DbgProvider/public/Debugger/DbgEngDebugger.cs [hunk 1]
```diff
@@ -1033,6 +1033,26 @@ internal bool QueryTargetIs32Bit()
...
 
+        internal bool QueryTargetIsWow64()
+        {
+            return ExecuteOnDbgEngThread( () =>
+            {
+                CheckHr(m_debugControl.GetActualProcessorType(out IMAGE_FILE_MACHINE type ));
+                if( type == IMAGE_FILE_MACHINE.AMD64 || (uint)type == 0xAA64 /*ARM64*/ ) //Wow64 supported architectures
+                {
+                    CheckHr(m_debugSystemObjects.GetCurrentThreadTeb( out var teb ));
+                    var wow64offset = teb + 0x1488; //TODO: TlsSlots[1] - 1 == WOW64_TLS_CPURESERVED
```
Comment (jazzdelightsme) [inline DbgProvider/public/Debugger/DbgEngDebugger.cs]: >teb + 0x1488 [](start = 38, length = 12)

This does not seem like a good idea. Even if the TEB size / layout never changed, what if the current process is an actual, real 64-bit process: in that case, somebody could be using that TLS slot, and we'll calculate the answer based on what is effectively random data.

The first idea that I had for answering the question "is it a wow64 target" would be to just check if wow64.dll is loaded. Anybody could name their binary wow64.dll, I guess, but you could also check the full path.

I asked around and somebody suggested that you could check if the currently executing code is wow by calling `GetActualProcessorType` and `GetExecutingProcessorType` (or potentially `GetEffectiveProcessorType` depending on what you’re doing). That won't work if the process is currently executing 64-bit code, but depending on why you are interested in wow-ness, maybe that's enough.

Another option might be to check the architecture of the main module, but I'm not sure off the top of my head how to do that.

I heard another suggestion to look at the `WowTebOffset` field, but it looks undocumented, and new to Win10, so I don't know if that'd actually be helpful.

File: DbgEngWrapper/DbgEngWrapper.h [hunk 2]
```diff
@@ -2767,6 +2767,12 @@ namespace DbgEngWrapper
             [In] ULONG BytesRequested,
             [In] BYTE* buffer,
             [Out] ULONG% BytesRead);
+		
+		generic <typename TValue>
```
Comment (jazzdelightsme) [inline DbgEngWrapper/DbgEngWrapper.h]: Could you please add a comment above this explaining that it's a convenience wrapper around `ReadVirtual`?

File: DbgProvider/public/Debugger/DbgEngDebugger.cs [hunk 3]
```diff
@@ -3189,7 +3304,7 @@ private void _EnsureSymbolsLoaded( DbgModuleInfo mod, CancellationToken cancelTo
                     if( mod.SymbolType == DEBUG_SYMTYPE.DEFERRED )
                     {
                         // TODO: progress output?
-                        CheckHr( m_debugSymbols.ReloadWide( " /f " + modFileName ) );
+                        CheckHr( DbgHelp.SymLoadModule( m_debugClient, mod.BaseAddress ) );
```
Comment (jazzdelightsme) [inline DbgProvider/public/Debugger/DbgEngDebugger.cs]: >SymLoadModule [](start = 41, length = 13)

I'm curious: why this change?

DbgShell bypasses dbgeng and goes to dbghelp to query type information, because we simply can't get enough type info from dbgeng. But in this case, I'm concerned that dbgeng may be maintaining some state, and by bypassing it and going directly to dbghelp to load a PDB, now dbgeng's state might be wrong.

File: DbgProvider/public/Debugger/DbgEngDebugger.cs [hunk 4]
```diff
@@ -2094,6 +2114,106 @@ public DbgModuleInfo GetModuleByAddress( ulong addrInModule )
             } );
         } // end GetModuleByName()
 
+        //Retrieves the ntdll module associated with what DbgEng considers to be the native architecture
```
Comment (jazzdelightsme) [inline DbgProvider/public/Debugger/DbgEngDebugger.cs]: >//Retrieves [](start = 8, length = 11)

Is there an .editorconfig setting for having a space after the opening "`//`"?

**Comment by Zhentar** [comment]
It's not so much that I *want* to detect Wow64 targets as it is that DbgEng forces me to detect Wow64 targets to get correct behaviors; regardless of the current effective machine type it only gives you the native PEB & TEB... so to get to the PEB&TEB associated with most of the code that does anything, you have to know that it's a Wow64 process, and then from there you have to pick from a set of undocumented pointer shenanigans a way to get to the 32 bit TEB.

It's possible there's some function somewhere in DbgEng that will just give you the 32 bit TEB/PEB, but nothing I found would do it.

File: DbgProvider/public/Debugger/DbgEngDebugger.cs (hunk 1 above)
Comment (Zhentar) [reply DbgProvider/public/Debugger/DbgEngDebugger.cs]: I borrowed this approach from wow64exts, so someone on the debugger team seems to have thought it was good enough... but I am inclined to agree it seems less than great.

DbgEng itself seems to have favored a variation on your approach - `ProcessInfo::MayBeWow64x86` checks if kernel32.dll is a folder named "syswow64".

WowTebOffset definitely doesn't do the trick; at least in Win 7 it's `SpareUlong0` and contains 0 for Wow64 processes (though I could check if the TIB ExceptionList is a pointer to the TEB address + 8192, but that doesn't feel like it's getting things any less janky)

The other idea I had was checking whether $ntwsym is populated, with a different value than $ntnsym. Still seems less than idea but at least it would be more directly tied to the assumptions being made about the implications of Wow64.

Checking the architecture of the main module seems like it could be a good idea... I think I'll take a look at that.

File: DbgProvider/public/Debugger/DbgEngDebugger.cs (hunk 3 above)
Comment (Zhentar) [reply DbgProvider/public/Debugger/DbgEngDebugger.cs]: Something had to be changed because `.reload /f ntdll.dll` doesn't tell DbgEng which ntdll.dll to load symbols for. I picked called DbgHelp directly over adding the base address & size to the reload call because the reload /f also causes DbgEng to send us ChangeSymbolState callbacks with no module for the argument.

DbgEng still knows about this happening - it's registered callbacks with DbgHelp, and we get a ChangeSymbolState callback from DbgEng before SymLoadModule has returned (and it actually passes along the right module base address, so we don't have to throw everything out!)

File: DbgProvider/public/Debugger/RealDebugEventCallbacks.cs [hunk 5]
```diff
@@ -564,22 +564,12 @@ public int ChangeSymbolState( DEBUG_CSS Flags, ulong Argument )
...
-
-                        // I wish we could refresh just the info for the module specified
-                        // by Argument, but we don't know what target this event is for,
-                        // and we can't call into dbgeng to find out.
-
-                        m_debugger.DiscardCachedModuleInfo();
+                        //If we load symbols using DbgEng's Reload command, it will send
+                        //us back an Argument of 0, and we'll be forced to throw away everything.
+                        //But in at least some other cases it faithfully passes along the 
+                        //base address - we still don't know which target it is for, but it is
```
Comment (jazzdelightsme) [inline DbgProvider/public/Debugger/RealDebugEventCallbacks.cs]: Oh, bypassing dbgeng and going to dbghelp's `SymLoadModuleEx` still causes the dbgeng ChangeSymbolState event to be raised? That would be interesting.

File: DbgProvider/public/Debugger/DbgEngDebugger.cs (hunk 4 above)
Comment (Zhentar) [reply DbgProvider/public/Debugger/DbgEngDebugger.cs]: It seems there is not. Resharper has added an impressive suite of editorconfig settings to their formatting, including angle bracket type parameter settings, so I've added those to the editorconfig, but while they have a `csharp_space_before_trailing_comment` for a space between code and the opening //, there's nothing for the right side.

File: DbgEngWrapper/DbgEngWrapper.h (hunk 2 above)
Comment (Zhentar) [reply DbgEngWrapper/DbgEngWrapper.h]: Added

File: DbgProvider/public/Debugger/DbgEngDebugger.cs (hunk 1 above)
Comment (Zhentar) [reply DbgProvider/public/Debugger/DbgEngDebugger.cs]: `GetExecutingProcessorType` definitely doesn't work for (Memory dumps of Wow64 processes from the Task Manager context menu option have an executing processor type of AMD64)

**Comment by Zhentar** [comment]
After mulling things over a bit, I realized that getting the effective architecture TEB is good enough for me, in which case comparing the effective architecture against the "actual" architecture works just fine.

Then the only thing we would need to know whether it's Wow64 or not for is to know whether or not getting the TEB address is a valid action (since you can switch the effective processor type to I386 even if it's a pure 64-bit process)... but that's enough of a strange edge case that I'm inclined to let it slide (particularly considering getting the TEB symbol will still throw)

File: DbgProvider/public/Debugger/RealDebugEventCallbacks.cs [hunk 6]
```diff
@@ -564,22 +564,12 @@ public int ChangeSymbolState( DEBUG_CSS Flags, ulong Argument )
...
-                        // the reason for the latter might be related to lots of extra
-                        // ".reload /f" from _EnsureSymbolsLoaded, which happened because
-                        // we hadn't properly refreshed the module info.
-
-                        // I wish we could refresh just the info for the module specified
-                        // by Argument, but we don't know what target this event is for,
-                        // and we can't call into dbgeng to find out.
-
-                        m_debugger.DiscardCachedModuleInfo();
+                        //If we load symbols using DbgEng's Reload command, it will send
```
Comment (jazzdelightsme) [inline DbgProvider/public/Debugger/RealDebugEventCallbacks.cs]: >// [](start = 24, length = 2)

nit: need space after `//`

**Comment by Zhentar** [comment]
There, think that's all of them :)
